File: petclinic-backend-ts/src/users/dto/user.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { IsBoolean, IsDefined, IsOptional, Length, ValidateNested } from 'class-validator';
import { RoleDto } from './role.dto';

/**
 * Data transfer object for a user.
 */
export class UserDto {
  @IsDefined({ message: 'must not be null' })
  @Length(1, 80, { message: 'size must be between 1 and 80' })
  @ApiProperty({ example: 'john.doe', description: 'The username' })
  username!: string;

  @IsOptional()
  @Length(1, 80, { message: 'size must be between 1 and 80' })
  @ApiPropertyOptional({ example: '1234abc', description: 'The password' })
  password?: string;

  @IsOptional()
  @IsBoolean()
  @ApiPropertyOptional({ example: true, description: 'Indicates if the user is enabled' })
  enabled?: boolean;

  @ValidateNested({ each: true })
  @Type(() => RoleDto)
  @ApiProperty({ type: () => [RoleDto], description: 'The roles of an user' })
  roles: RoleDto[] = [];
}

File: petclinic-backend/src/main/java/victor/training/petclinic/rest/dto/UserDto.java
package victor.training.petclinic.rest.dto;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import org.springframework.lang.Nullable;

import java.util.ArrayList;
import java.util.List;

public class UserDto {

    @NotNull
    @Size(min = 1, max = 80)
    @Schema(example = "john.doe", description = "The username")
    private String username;

    @Size(min = 1, max = 80)
    @Schema(example = "1234abc", description = "The password")
    private @Nullable String password;

    @Schema(example = "true", description = "Indicates if the user is enabled")
    private @Nullable Boolean enabled;

    @Valid
    @Schema(description = "The roles of an user")
    private List<@Valid RoleDto> roles = new ArrayList<>();

    public String getUsername() {
        return username;
    }

    public UserDto setUsername(String username) {
        this.username = username;
        return this;
    }

    public String getPassword() {
        return password;
    }

    public UserDto setPassword(String password) {
        this.password = password;
        return this;
    }

    public Boolean getEnabled() {
        return enabled;
    }

    public UserDto setEnabled(Boolean enabled) {
        this.enabled = enabled;
        return this;
    }

    public List<RoleDto> getRoles() {
        return roles;
    }

    public UserDto setRoles(List<RoleDto> roles) {
        this.roles = roles;
        return this;
    }
}
